Mathlib/Algebra/Module/SnakeLemma.lean

```diff
@@ -0,0 +1,223 @@
+        ι₂      ι₃
+        ↓       ↓
+M₁ -f₁→ M₂ -f₂→ M₃
+|       |       |
+i₁      i₂      i₃
+↓       ↓       ↓
+N₁ -g₁→ N₂ -g₂→ N₃
+|
+π₁
+↓
+C₁
+
+```
+Such that `f₂` is surjective with a (set-theoretic) section `ρ`, `g₁` is injective with a
+(set-theoretic) retraction `π`, and `ι₃` is injective, then `K₂ -F→ K₂ -δ→ C₁` is exact.
+-/
+lemma SnakeLemma.exact_δ_right (F : K₂ →ₗ[R] K₃) (hF : f₂.comp ι₂ = ι₃.comp F)
+    (h : Injective ι₃) :
+    Exact F (δ i₁ i₂ i₃ f₁ f₂ hf g₁ g₂ hg h₁ h₂ σ hσ ρ hρ ι₃ hι₃ π₁ hπ₁) := by
+  haveI H₁ : ∀ x, f₂ (σ x) = x := congr_fun hσ
+  haveI H₂ := δ_aux i₂ i₃ f₂ g₁ g₂ hg h₂ σ hσ ρ hρ ι₃ hι₃
+  intro x
+  constructor
+  · intro H
+    obtain ⟨y, hy⟩ := (hπ₁ _).mp H
+    obtain ⟨k, hk⟩ : σ (ι₃ x) - f₁ y ∈ Set.range ι₂ := by
+      rw [← hι₂, map_sub, ← H₂, ← hy, sub_eq_zero]; exact congr($h₁ y)
+    refine ⟨k, h ?_⟩
+    rw [← ι₃.comp_apply, ← hF, f₂.comp_apply, hk, map_sub, H₁, hf.apply_apply_eq_zero, sub_zero]
+  · rintro ⟨y, rfl⟩
+    exact (δ_eq i₁ i₂ i₃ f₁ f₂ hf g₁ g₂ hg h₁ h₂ σ hσ ρ hρ ι₃ hι₃ π₁ hπ₁ _ (ι₂ y) congr($hF y)
+      _ (by rw [map_zero, hι₂.apply_apply_eq_zero])).trans π₁.map_zero
+
+include hπ₂ in
+/--
+Supppose we have an exact commutative diagram
+```
+                K₃
+                |
+                ι₃
+                ↓
+M₁ -f₁→ M₂ -f₂→ M₃
+|       |       |
+i₁      i₂      i₃
+↓       ↓       ↓
+N₁ -g₁→ N₂ -g₂→ N₃
+|       |
+π₁      π₂
+↓       ↓
+C₁ -G-→ C₂
+
+```
+Such that `f₂` is surjective with a (set-theoretic) section `ρ`, `g₁` is injective with a
+(set-theoretic) retraction `π`, and `π₁` is surjective, then `K₂ -δ→ C₁ -G→ C₂` is exact.
+-/
+lemma SnakeLemma.exact_δ_left (G : C₁ →ₗ[R] C₂) (hF : G.comp π₁ = π₂.comp g₁) (h : Surjective π₁) :
+    Exact (δ i₁ i₂ i₃ f₁ f₂ hf g₁ g₂ hg h₁ h₂ σ hσ ρ hρ ι₃ hι₃ π₁ hπ₁) G := by
+  haveI H₁ : ∀ x, f₂ (σ x) = x := congr_fun hσ
+  haveI H₂ := δ_aux i₂ i₃ f₂ g₁ g₂ hg h₂ σ hσ ρ hρ ι₃ hι₃
+  intro x
+  constructor
+  · intro H
+    obtain ⟨x, rfl⟩ := h x
+    obtain ⟨y, hy⟩ := (hπ₂ (g₁ x)).mp (by simpa only [← LinearMap.comp_apply, hF] using H)
+    obtain ⟨z, hz⟩ : f₂ y ∈ range ι₃ := (hι₃ (f₂ y)).mp (by rw [← i₃.comp_apply, ← h₂,
+      g₂.comp_apply, hy, hg.apply_apply_eq_zero])
+    exact ⟨z, δ_eq i₁ i₂ i₃ f₁ f₂ hf g₁ g₂ hg h₁ h₂ σ hσ ρ hρ ι₃ hι₃ π₁ hπ₁ _ _ hz.symm _ hy.symm⟩
+  · rintro ⟨x, rfl⟩
+    simp only [δ, id_eq, coe_mk, AddHom.coe_mk]
+    rw [← G.comp_apply, hF, π₂.comp_apply, H₂, hπ₂.apply_apply_eq_zero]
+
+/--
+Supppose we have an exact commutative diagram
+```
+                K₃
+                |
+                ι₃
+                ↓
+M₁ -f₁→ M₂ -f₂→ M₃
+|       |       |
+i₁      i₂      i₃
+↓       ↓       ↓
+N₁ -g₁→ N₂ -g₂→ N₃
+|
+π₁
+↓
+C₁
+
+```
+Such that `f₂` is surjective and `g₁` is injective,
+then this is the linear map `K₃ → C₁` given by the snake lemma.
+Also see `SnakeLemma.δ` for a noncomputable version.
+-/
+noncomputable def SnakeLemma.δ' (hf₂ : Surjective f₂) (hg₁ : Injective g₁) : K₃ →ₗ[R] C₁ :=
```

Comment:
I think it would be good to restate `δ_eq`, `exact_δ_right` and `exact_δ_left` for `δ'`.